### ISSUE TITLE
VkPhysicalDeviceFeatures2 & vkCreateDevice clarification request

### ISSUE THREAD

**Issue by darksylinc** [KhronosGroup/Vulkan-Docs#1743]
In https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPhysicalDeviceFeatures2.html it says:

> The pNext chain of this structure is used to extend the structure with features defined by extensions. This structure **can** be used in vkGetPhysicalDeviceFeatures2 or **can** be included in the pNext chain of a VkDeviceCreateInfo structure, in which case it controls which features are enabled in the device in lieu of pEnabledFeatures.

Note that it says **can**, not **must**.

However I found out [the hard way](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/3688) that `VkPhysicalDeviceFeatures2` **must** be passed to `VkDeviceCreateInfo::pNext` in order for the validation layers to properly account of these settings (and the same goes for each extension, e.g. `VkPhysicalDevice16BitStorageFeatures` and `VkPhysicalDeviceShaderFloat16Int8Features` need to be passed as well).

The way the wording says, implies that if I don't pass `VkPhysicalDeviceFeatures2` to `VkDeviceCreateInfo::pNext` then defaults will be used.

However the validation layers _require_ it, otherwise the validation will get all 0s in these structs.

As far as I can gather, either the validation layers have a bug or the documentation should more clearly state it must be passed along. IMO this information should be included in [vkCreateDevice](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCreateDevice.html) or [VkDeviceCreateInfo](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDeviceCreateInfo.html) because it pertains many extensions, not just VkPhysicalDeviceFeatures2.

The only relevant bit of information I could find was in [Vulkan Memory Allocator](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/vk_amd_device_coherent_memory.html) which asks to fill **both**  ppEnabledExtensionNames  and `VkDeviceCreateInfo::pNext`

## Update

Finally I found the [relevant section "Features"](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#features):

> Setting pEnabledFeatures to NULL and not including a VkPhysicalDeviceFeatures2 in the pNext chain of VkDeviceCreateInfo is equivalent to setting all members of the structure to VK_FALSE.

I believe IMHO the documentation should be far more clear on this. Right now it's very obscured away.

In other words this is what happened: why the validation layer says "VK_KHR_16bit_storage" is not enabled even though it's clearly part of createInfo.ppEnabledExtensionNames?
 - Because I didn't pass VkPhysicalDevice16BitStorageFeatures to VkDeviceCreateInfo::pNext. This is not obvious at all.

**Comment by sfricke-samsung** [comment]
@darksylinc First thanks for taking time to write up all the details here and the original VVL issue

I am aware of the complex nature of the features and those promoted to core

1. We have a Vulkan Guide page on [enabling extensions](https://github.com/KhronosGroup/Vulkan-Guide/blob/master/chapters/enabling_features.adoc) which we wrote to be used as mapping all the parts of the spec can be annoying

2. As you pointed out

> This structure **can** be used in vkGetPhysicalDeviceFeatures2 or **can** be included in the pNext chain of a VkDeviceCreateInfo structure

The reason it is not **must** is it is very valid for a developer to query with `vkGetPhysicalDeviceFeatures2`, find the features it wanted were not supported and then **not bother** adding it to `VkDeviceCreateInfo`. By changing it to **must** it would imply you **must** include it even if you don't need the features

3. I agree the VVL error message was not the best, will take action to fix that! https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/3699

4. For your

> As far as I can gather, either the validation layers have a bug or the documentation should more clearly state it must be passed along.

I believe the VVL don't have a bug and also the Vulkan Spec states everything need, it is not easy to read, but that is why we created the Vulkan Guide entry, so if anything I would argue that should be improved instead if anything there is unclear

**Comment by sfricke-samsung** [comment]
@darksylinc going to close the issue as explained the spec contains the information and the Vulkan Guide has a chapter to help developers. Feel free to re-open if you or anyone has a purpose suggestion of what wording to change in the spec

**Comment by darksylinc** [comment]
I've had time to think on this.

 * I'm not against the design choices.
 * As you pointed out the spec wording is technically correct

I just feel the tutorials and perhaps `vkCreateDevice` should highlight more that you should query VkPhysicalDeviceFeatures2 and pass it when you're using this extension.

The context where I was coming from is I already had Vulkan initialization code setup 12 months ago and the details were a bit hazy when I had to add the extension. Thus highlighting this important detail can help troubleshoot problems.
This is specially true when you're using extensions.

Most tutorials stick to core initialization so this step is often skipped and details are lost.
